File: WoofIOS/SwiftPackages/DesignSystem/Sources/DesignSystem/Atomic/Organisms/Dialog/OrgAlertDialog.swift
import SwiftUI

public struct OrgAlertDialog: ViewModifier {
    private let isPresented: Bool
    private let title: String
    private let message: String
    private let confirmText: String
    private let dismissText: String
    private let onConfirmTap: () -> Void
    private let onDismissTap: () -> Void
    private let isDestructive: Bool

    public init(
        isPresented: Bool,
        title: String,
        message: String,
        confirmText: String,
        dismissText: String,
        onConfirmTap: @escaping () -> Void,
        onDismissTap: @escaping () -> Void,
        isDestructive: Bool = false
    ) {
        self.isPresented = isPresented
        self.title = title
        self.message = message
        self.confirmText = confirmText
        self.dismissText = dismissText
        self.onConfirmTap = onConfirmTap
        self.onDismissTap = onDismissTap
        self.isDestructive = isDestructive
    }

    public func body(content: Content) -> some View {
        content
            .alert(title, isPresented: Binding(get: { isPresented }, set: { _ in })) {
                Button(dismissText, role: .cancel, action: onDismissTap)
                Button(confirmText, role: isDestructive ? .destructive : nil, action: onConfirmTap)
            } message: {
                Text(message)
            }
    }
}

public extension View {
    func orgAlertDialog(
        isPresented: Bool,
        title: String,
        message: String,
        confirmText: String,
        dismissText: String,
        onConfirmTap: @escaping () -> Void,
        onDismissTap: @escaping () -> Void,
        isDestructive: Bool = false
    ) -> some View {
        modifier(
            OrgAlertDialog(
                isPresented: isPresented,
                title: title,
                message: message,
                confirmText: confirmText,
                dismissText: dismissText,
                onConfirmTap: onConfirmTap,
                onDismissTap: onDismissTap,
                isDestructive: isDestructive
            )
        )
    }
}

File: WoofAndroid/design-system/src/main/kotlin/com/perrystreet/woof/designsystem/atomic/organisms/dialog/OrgAlertDialog.kt
package com.perrystreet.woof.designsystem.atomic.organisms.dialog

import androidx.compose.material3.AlertDialog
import androidx.compose.runtime.Composable
import com.perrystreet.woof.designsystem.atomic.atoms.text.AtomText
import com.perrystreet.woof.designsystem.atomic.atoms.text.roles.TextColorRole
import com.perrystreet.woof.designsystem.atomic.atoms.text.roles.TextFontRole
import com.perrystreet.woof.designsystem.atomic.molecules.button.MolButtonCompact
import com.perrystreet.woof.designsystem.atomic.molecules.button.roles.ButtonRole
import com.perrystreet.woof.designsystem.theme.Theme

@Composable
fun OrgAlertDialog(
    title: String,
    message: String,
    confirmText: String,
    dismissText: String,
    onConfirmTap: () -> Unit,
    onDismissTap: () -> Unit,
    isDestructive: Boolean = false,
) {
    val confirmRole = when (isDestructive) {
        true -> ButtonRole.Destructive
        false -> ButtonRole.Primary
    }
    AlertDialog(
        onDismissRequest = onDismissTap,
        containerColor = Theme.colors.surfaceContainer,
        title = { AtomText(text = title, textFontRole = TextFontRole.DisplayH3) },
        text = {
            AtomText(
                text = message,
                textFontRole = TextFontRole.BodyP1,
                colorRole = TextColorRole.OnSurfaceVariant,
            )
        },
        confirmButton = { MolButtonCompact(text = confirmText, onTap = onConfirmTap, role = confirmRole) },
        dismissButton = { MolButtonCompact(text = dismissText, onTap = onDismissTap, role = ButtonRole.Secondary) },
    )
}
